onnx/defs/tensor/defs.cc

```diff
@@ -1185,4 +1185,122 @@ ONNX_OPERATOR_SET_SCHEMA(
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    OneHot,
+    9,
+    OpSchema()
+        .SetDoc(OneHot_ver9_doc)
+        .Attr(
+            "axis",
+            "(Optional) Axis along which one-hot representation in added. Default: axis=-1.",
+            AttributeProto::INT,
+            static_cast<int64_t>(-1))
+        .Input(
+            0,
+            "indices",
+            "Input tensor containing indices. The values must be non-negative integers. "
+            "Any entries in the 'indices' input tensor with values outside the range [0, depth) "
+            "will result in one-hot representation with all 'off_value' values in the output tensor."
+            "In case 'indices' is of non-integer type, the values will be casted to int64 before use.",
+            "T1")
+        .Input(
+            1,
+            "depth",
+            "Scalar specifying the number of classes in one-hot tensor. This is also the size "
+            "of the one-hot dimension (specified by 'axis' attribute) added on in the output "
+            "tensor and the values in the 'indices' input tensor are expected to be "
+            "in the range [0, depth). The"
+            "In case 'depth' is of non-integer type, it will be casted to int64 before use.",
+            "T1")
+        .Input(
+            2,
+            "values",
+            "Rank 1 tensor containing exactly two elements, in the format [off_value, on_value], "
+            "where 'on_value' is the value used for filling locations specified in 'indices' input "
+            "tensor, and 'off_value' is the value used for filling locations other than those specified "
+            "in 'indices' input tensor. ",
+            "T2")
+        .Output(0,
+            "output",
+            "Tensor of rank one greater than input tensor 'indices', i.e. rank(output) = rank(indices) + 1. "
+            "The data type for the elements of the output tensor is the same as the type of input 'values' "
+            "is used.",
+            "T2")
+        .TypeConstraint(
+            "T1",
+            OpSchema::all_numeric_types(),
+            "Constrains input to only numeric types.")
+        .TypeConstraint(
+            "T2",
+            OpSchema::all_tensor_types(),
+            "Constrain to any tensor type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          // Check that the node has two or three onputs.
+          if (!hasNInputShapes(ctx, 3)) {
+            fail_shape_inference(
+                  "OneHot node must have three inputs.");
+          }
+          // If input 'values' is specified, it must be a two-element vector.          
+          auto& input_shape = getInputShape(ctx, 2);
+          if (input_shape.dim_size() != 1) {
+            fail_type_inference(
+                  "Input 'values' must be rank 1 tensor.");
+          }
+          if (!input_shape.dim((int)0).has_dim_value() ||
+            (input_shape.dim((int)0).has_dim_value() &&
```

Comment:
ditto, if we don't have enough information, just exit the shape inference.

Reply:
Agreed. Updated shape inference code at multiple places to do that.